File: CALCULATOR/Calculator.java

public class Calculator {

    private String expression = "";

    // ================= BASIC METHODS =================

    public void appendNumber(int num) {
        expression += num;
    }

    public void appendDecimal() {
        if (!expression.endsWith(".") &&
                (expression.isEmpty() ||
                 Character.isDigit(expression.charAt(expression.length() - 1)))) {
            expression += ".";
        }
    }

    public void appendOperator(char operator) {
        if (!expression.isEmpty() &&
                !isOperator(expression.charAt(expression.length() - 1))) {
            expression += " " + operator + " ";
        }
    }

    private boolean isOperator(char ch) {
        return ch == '+' || ch == '-' || ch == '*' || ch == '/';
    }

    public void deleteLast() {
        if (!expression.isEmpty()) {
            expression = expression.substring(0, expression.length() - 1);
        }
    }

    public void clear() {
        expression = "";
    }

    public String getExpression() {
        return expression.isEmpty() ? "0" : expression;
    }

    // ================= CALCULATION =================

    public double calculate() throws Exception {
        if (expression.isEmpty()) return 0;
        return evaluateExpression(expression);
    }

    private double evaluateExpression(String expr) throws Exception {

        expr = expr.replaceAll("\\s+", "");

        // First handle multiplication & division
        expr = handleMultiplicationDivision(expr);

        // Then handle addition & subtraction
        expr = handleAdditionSubtraction(expr);

        return Double.parseDouble(expr);
    }

    private String handleMultiplicationDivision(String expr) throws Exception {

        while (expr.contains("*") || expr.contains("/")) {

            int mulIndex = expr.indexOf("*");
            int divIndex = expr.indexOf("/");

            int opIndex;
            char op;

            if (mulIndex == -1) {
                opIndex = divIndex;
                op = '/';
            } else if (divIndex == -1) {
                opIndex = mulIndex;
                op = '*';
            } else {
                if (mulIndex < divIndex) {
                    opIndex = mulIndex;
                    op = '*';
                } else {
                    opIndex = divIndex;
                    op = '/';
                }
            }

            // LEFT OPERAND
            int leftStart = opIndex - 1;
            while (leftStart > 0 &&
                    (Character.isDigit(expr.charAt(leftStart - 1)) ||
                     expr.charAt(leftStart - 1) == '.')) {
                leftStart--;
            }

            double left = Double.parseDouble(expr.substring(leftStart, opIndex));

            // RIGHT OPERAND
            int rightStart = opIndex + 1;

            if (rightStart < expr.length() &&
                    expr.charAt(rightStart) == '-') {
                rightStart++;
            }

            int rightEnd = rightStart;

            while (rightEnd < expr.length() &&
                    (Character.isDigit(expr.charAt(rightEnd)) ||
                     expr.charAt(rightEnd) == '.')) {
                rightEnd++;
            }

            double right = Double.parseDouble(expr.substring(opIndex + 1, rightEnd));

            double result;

            if (op == '*') {
                result = left * right;
            } else {
                if (right == 0) {
                    throw new Exception("Division by zero");
                }
                result = left / right;
            }

            expr = expr.substring(0, leftStart) +
                   result +
                   expr.substring(rightEnd);
        }

        return expr;
    }

    private String handleAdditionSubtraction(String expr) {

        if (expr.startsWith("-")) {
            expr = "0" + expr;
        }

        double accumulator = 0;
        double currentNumber = 0;
        char lastOperator = '+';

        for (int i = 0; i < expr.length(); i++) {

            char ch = expr.charAt(i);

            if (Character.isDigit(ch) || ch == '.') {

                int start = i;
                while (i < expr.length() &&
                        (Character.isDigit(expr.charAt(i)) ||
                         expr.charAt(i) == '.')) {
                    i++;
                }

                currentNumber = Double.parseDouble(expr.substring(start, i));
                i--;

            } else if (ch == '+' || ch == '-') {

                if (lastOperator == '+') {
                    accumulator += currentNumber;
                } else {
                    accumulator -= currentNumber;
                }

                lastOperator = ch;
                currentNumber = 0;
            }
        }

        if (lastOperator == '+') {
            accumulator += currentNumber;
        } else {
            accumulator -= currentNumber;
        }

        return String.valueOf(accumulator);
    }

    // ================= EXTRA FEATURES =================

    public double findSquare(double number) {
        return number * number;
    }

    public double findSquareRoot(double number) throws Exception {
        if (number < 0)
            throw new Exception("Cannot calculate square root of negative number");
        return Math.sqrt(number);
    }

    public String decimalToFraction(double decimal) {

        if (decimal == (long) decimal)
            return String.format("%d/1", (long) decimal);

        double tolerance = 1e-10;
        long numerator = 1;
        long denominator = 1;

        while (Math.abs(decimal - (double) numerator / denominator) > tolerance) {
            numerator++;
            denominator = (long) Math.round(numerator / decimal);
        }

        long gcd = gcd(numerator, denominator);
        return (numerator / gcd) + "/" + (denominator / gcd);
    }

    public double fractionToDecimal(int numerator, int denominator) throws Exception {
        if (denominator == 0)
            throw new Exception("Denominator cannot be zero");
        return (double) numerator / denominator;
    }

    private long gcd(long a, long b) {
        return b == 0 ? a : gcd(b, a % b);
    }

    // ================= MAIN TEST =================

    public static void main(String[] args) throws Exception {

        Calculator calc = new Calculator();

        calc.appendNumber(5);
        calc.appendOperator('+');
        calc.appendNumber(3);

        System.out.println("Expression: " + calc.getExpression());
        System.out.println("Result: " + calc.calculate());

        calc.clear();
        calc.appendNumber(2);
        calc.appendOperator('*');
        calc.appendNumber(3);
        calc.appendOperator('+');
        calc.appendNumber(4);

        System.out.println("Expression: " + calc.getExpression());
        System.out.println("Result: " + calc.calculate());

        System.out.println("Square of 7: " + calc.findSquare(7));
        System.out.println("Square root of 16: " + calc.findSquareRoot(16));
        System.out.println("0.75 as fraction: " + calc.decimalToFraction(0.75));
        System.out.println("3/4 as decimal: " + calc.fractionToDecimal(3, 4));
    }
}

File: CALCULATOR/calculator.js
let display = document.getElementById("display");
let expression = "";

let recognition = null;
let isListening = false;
let isWakeWordMode = true;
let lastTranscript = "";

/* ================= VOICE INIT ================= */

/* ================= VOICE INIT ================= */

function initSpeechRecognition() {
    const SpeechRecognition =
        window.SpeechRecognition || window.webkitSpeechRecognition;

    if (!SpeechRecognition) {
        alert("Use Chrome or Edge browser.");
        return;
    }

    recognition = new SpeechRecognition();
    recognition.continuous = true;
    recognition.interimResults = false;
    recognition.lang = "en-US";

    recognition.onresult = (event) => {
        const transcript =
            event.results[event.results.length - 1][0].transcript
                .toLowerCase()
                .trim();

        console.log("Voice:", transcript);

        // WAKE WORD MODE
        if (isWakeWordMode) {
            if (/\b(hey|hi|hello|ok)\s+(calculator|calc)\b/.test(transcript)) {
                isWakeWordMode = false;
                speakResult("Calculator activated");
                updateVoiceStatus("Listening for command...");
                return;
            }
        }

        // COMMAND MODE
        if (!isWakeWordMode) {
            processVoiceCommand(transcript);
        }
    };

    recognition.onerror = (e) => {
        console.log("Mic error:", e.error);
    };

    recognition.onend = () => {
        // Keep listening continuously unless explicitly stopped by closing page
        try {
            recognition.start();
        } catch(e) {}
    };
}




function updateVoiceStatus(text) {
    const el = document.getElementById("voice-status");
    if (el) el.textContent = text;
}

/* ================= VOICE PROCESS ================= */

function processVoiceCommand(transcript) {
    if (transcript.includes("stop") || transcript.includes("deactivate")) {
        isWakeWordMode = true;
        speakResult("Calculator deactivated");
        updateVoiceStatus('Waiting for "Hey Calculator"...');
        return;
    }

    if (transcript.includes("clear")) {
        clearDisplay();
        speakResult("Cleared");
        return;
    }

   // ===== Square Root Handling =====
if (transcript.includes("square root") || transcript.includes("root")) {

    let number = extractNumber(transcript);
    if (number !== null) {
        let result = Math.sqrt(number);
        expression = result.toString();
        updateDisplay();
        speakResult("Square root is " + result);
        return;
    }

    return findSquareRoot();
}

// ===== Square Handling =====
if (transcript.includes("square")) {

    let number = extractNumber(transcript);
    if (number !== null) {
        let result = number * number;
        expression = result.toString();
        updateDisplay();
        speakResult("Square is " + result);
        return;
    }

    return findSquare();
}
    if (transcript.includes("decimal to fraction")) return decimalToFraction();
    if (transcript.includes("fraction to decimal")) return fractionToDecimal();

    parseMath(transcript);
}

/* ================= NUMBER PARSER ================= */

function parseMath(transcript) {

    // 🔥 NEW: If transcript already contains numbers and operators
    if (/^[0-9+\-*/.\s]+$/.test(transcript)) {
        expression = transcript.replace(/\s+/g, "");
        updateDisplay();
        calculate();
        return;
    }

    // 🔽 Old word-based logic continues below

    transcript = transcript.replace(
        /(twenty|thirty|forty|fifty|sixty|seventy|eighty|ninety)(one|two|three|four|five|six|seven|eight|nine)/g,
        "$1 $2"
    );

    const units = {
        zero: 0, one: 1, two: 2, three: 3, four: 4,
        five: 5, six: 6, seven: 7, eight: 8, nine: 9
    };

    const teens = {
        ten: 10, eleven: 11, twelve: 12, thirteen: 13,
        fourteen: 14, fifteen: 15, sixteen: 16,
        seventeen: 17, eighteen: 18, nineteen: 19
    };

    const tens = {
        twenty: 20, thirty: 30, forty: 40,
        fifty: 50, sixty: 60, seventy: 70,
        eighty: 80, ninety: 90
    };

    let words = transcript.split(/\s+/);
    let result = "";
    let i = 0;

    while (i < words.length) {
        let word = words[i];

        if (teens[word] !== undefined) {
            result += teens[word];
            i++;
            continue;
        }

        if (tens[word] !== undefined) {
            let value = tens[word];
            if (units[words[i + 1]] !== undefined) {
                value += units[words[i + 1]];
                i++;
            }
            result += value;
            i++;
            continue;
        }

        if (units[word] !== undefined) {
            result += units[word];
            i++;
            continue;
        }

        if (word.match(/plus|add|pulse/)) result += "+";
        else if (word.match(/minus|subtract/)) result += "-";
        else if (word.match(/times|multiply|into/)) result += "*";
        else if (word.match(/divide|divided|by/)) result += "/";
        else if (word === "point" || word === "dot") result += ".";

        i++;
    }

    if (!result) return;

    // SMART APPEND LOGIC
if (expression && /^[+\-*/]/.test(result)) {
    expression += result;
} else {
    expression = result;
}
    updateDisplay();

    if (/^[0-9+\-*/.]+$/.test(expression)) {
        calculate();
    }
}
/* ================= CALCULATOR ================= */

function appendNumber(num) {
    expression = expression === "0" ? num : expression + num;
    updateDisplay();
}

function appendOperator(op) {
    if (expression === "" && op !== "-") return;

    if (/[+\-*/]$/.test(expression)) {
        expression = expression.slice(0, -1) + op;
    } else {
        expression += op;
    }
    updateDisplay();
}

function updateDisplay() {
    display.value = expression || "0";
}

function clearDisplay() {
    expression = "";
    updateDisplay();
}

function deleteLast() {
    expression = expression.slice(0, -1);
    updateDisplay();
}

function calculate() {
    if (!expression) return;

    try {
        if (!/^[0-9+\-*/.]+$/.test(expression))
            throw new Error("Invalid");

        let result = eval(expression);
        if (!isFinite(result)) throw new Error("Math Error");

        result = Math.round(result * 100000000) / 100000000;
        expression = result.toString();
        updateDisplay();
        speakResult("The result is " + result);
    } catch {
        display.value = "Error";
        expression = "";
    }
}

/* ================= EXTRA FEATURES ================= */

function findSquare() {
    let num = parseFloat(expression);
    if (isNaN(num)) return;
    let result = num * num;
    expression = result.toString();
    updateDisplay();
    speakResult("Square is " + result);
}

function findSquareRoot() {
    let num = parseFloat(expression);
    if (isNaN(num) || num < 0) {
        display.value = "Error";
        expression = "";
        return;
    }
    let result = Math.sqrt(num);
    expression = result.toString();
    updateDisplay();
    speakResult("Square root is " + result);
}

function decimalToFraction() {
    let decimal = parseFloat(expression);
    if (isNaN(decimal)) return;

    let denominator = 1000000;
    let numerator = Math.round(decimal * denominator);
    let gcdVal = gcd(numerator, denominator);

    expression = numerator / gcdVal + "/" + denominator / gcdVal;
    updateDisplay();
}

function fractionToDecimal() {
    if (!expression.includes("/")) return;

    let [n, d] = expression.split("/");
    let result = parseFloat(n) / parseFloat(d);

    expression = result.toString();
    updateDisplay();
}

function gcd(a, b) {
    return b === 0 ? a : gcd(b, a % b);
}

function speakResult(text) {
    if ("speechSynthesis" in window) {
        let utter = new SpeechSynthesisUtterance(text);
        utter.rate = 0.9;
        speechSynthesis.speak(utter);
    }
}
function extractNumber(text) {
    let match = text.match(/\d+(\.\d+)?/);
    return match ? parseFloat(match[0]) : null;
}

/* ================= INIT ================= */

document.addEventListener("DOMContentLoaded", () => {
    initSpeechRecognition();
    if (recognition) {
        try {
            recognition.start();
            isListening = true;
            updateVoiceStatus('Waiting for "Hey Calculator"...');
        } catch (e) {
            updateVoiceStatus('Click anywhere to enable mic initially...');
            // Fallback for strict browsers requiring interact to start mic
            document.body.addEventListener('click', () => {
                try {
                    recognition.start();
                    updateVoiceStatus('Waiting for "Hey Calculator"...');
                } catch(err) {}
            }, {once: true});
        }
    }
});